### PR TITLE
Update build system to work on Windows et al, upgrade test_*.py

### PR DESCRIPTION
**Pull request by scivision**
* Upgraded setup.cfg, setup.py to current (2016) standards, greatly simplifying setup process and code across Windows, Linux, Mac, etc.
*  updated test syntax, to greatly dedupe test code by using pytest.mark.parametrize
* put code  to pep8 standard

Note: the develop branch is broken for Windows due to the ctypes code recently added.  This new Master / Develop2 does work on Windows, Linux etc.

### CODE REVIEW

**Comment by aburrell** [comment]
@scivision thanks for putting in some development work.  Please recall that all pull requests must be made to the develop branch, not master.

**Comment by scivision** [comment]
The develop branch is going the wrong way--it is not platform independent. I will have to make this my own project then, we need it for our work. A better path is probably to figure out why develop is breaking rather than have competing projects. This is more than "some work"

**Comment by scivision** [comment]
For those who wish to have this program work on Windows as well as Linux, the ApexPy fork is being developed at https://github.com/scivision/apexpy

**Comment by aburrell** [comment]
If you are not prepared to work within the develop branch, which Ashton has done some significant work on, then it is best that you create a different project.

File: setup.py [hunk 1]
```diff
@@ -1,92 +1,28 @@
...
 
 # Include extensions only when not on readthedocs.org
-if environ.get('READTHEDOCS', None) == 'True':
+if os.environ.get('READTHEDOCS', None) == 'True':
     from setuptools import setup
     from distutils.core import Extension
     extensions = []
 else:
     from numpy.distutils.core import setup, Extension
+
```
Comment (asreimer) [inline setup.py]: Apexpy worked fine on windows before, why is this needed suddenly?

**Comment by asreimer** [comment]
@scivision We are not intentionally breaking windows support. AFAIK the current released version of apexpy have worked fine on windows, so just because the unreleased development branch doesn't work right now doesn't mean we want that to be the case. In fact, @aburrell has been soliciting help from people who use windows to get this figured out.

Do you have any insight as to why ctypes is breaking the windows side of things? I haven't had a chance to do that testing myself yet and you seem to have spent a lot of time working in windows with this code. A quick glance at the changes you are proposing here

@aburrell and @scivision, in a community that has enough fragmented code/projects, it would be much better if we could work together on things instead of continuing to go our own way. @scivision, I doubt very much that @aburrell meant to demean your contribution here and instead was genuinely appreciative of your contribution.

I can make an effort to help. I am finally getting anaconda and such installed on my home windows computer. I just need some direction.

File: setup.py [hunk 2]
```diff
@@ -1,92 +1,28 @@
...
+    # add MinGW to Windows only, if needed
+    if os.name == 'nt':
+        sfn = os.path.join(os.path.dirname(__file__), 'setup.cfg')
+        stxt = open(sfn).read()
+        if '[build_ext]' not in stxt:
+            with open(sfn, 'a') as f:
+                f.write("\n[build_ext]\ncompiler = mingw32")
```
Comment (asreimer) [inline setup.py]: This seems to hardcode in the mingw32 compiler. What about mingw64, etc? There's got to be a better way to detect the appropriate compiler on windows? Maybe not.

Comment (scivision) [reply setup.py]: This is a workaround for Numpy distutils behavior with MinGW on Windows, it is a longstanding issue (many years). Numpy 1.16 introduced new functionality I am using in other projects. I didn't want to introduce a Numpy &ge; 1.16 requirement at this time.

Comment (scivision) [reply setup.py]: That is, with Numpy &ge; 1.16 this setup.py blurb is not needed. My Fortran + Python projects have a one-liner setup.py just like plain Python projects:
```python
import setuptools; setuptools.setup()
```
and everything else is in setup.cfg and other static files (mypy.ini, .flake8, .coveragerc, etc.)

Comment (scivision) [reply setup.py]: `mingw32` actually is also for MinGW64, which is what I normally use for Gfortran on native Windows

Comment (asreimer) [reply setup.py]: Thanks for the info! This will help when I dig in with my windows machine.

**Comment by scivision** [comment]
We just had to get some work (publications, proposals) out on other projects, and this code was being used internally. I have had analogous situation before, where I needed to make even more extreme--actual architectural changes to make a major Python project more flexible. I ran my fork for a year or so, and the original project maintainers nibbled pieces from my fork back in until they could converge again.

**Comment by scivision** [comment]
I think we'd consider to rearchitect:

* how the CI is setup, there may be a more streamlined way to do this. I do CI across Python versions, compiler vendor/version, Python implementation (CPython vs. PyPy) and operating system with significantly smaller / simpler test config files.
* how the Fortran integration is done, at the same time as updating the metadata.

For now I may rely on my ApexPy fork as this works in these regards, less CI (tests pass on local computers regardless of operating system). However I can introduce a series of PRs to bring these upgrades here as well. These changes are perhaps best addressed in a series of PRs due to their scope and size.

File: setup.py (hunk 1 above)
Comment (scivision) [reply setup.py]: The install silently fails on windows.
